### ISSUE TITLE
False negative for NameError from calling inner function before variable is assigned

### ISSUE THREAD

**Issue by jacobtylerwalls** [PyCQA/pylint#8495]
### Bug description

```python
def outer():
    def inner():
        print(i)
    
    inner()
    for i in range(5):
        inner()

outer()
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
no msg
```

### Expected behavior

A new message possibly-used-before-assignment on line 3. Notice the second call to `inner()` executes fine.

### Pylint version

```shell
pylint 3.0.0a6
astroid 2.16.0dev0
Python 3.11.2
```
